src/exit_handler/src/exit_handler_intel_x64_eapis_cpuid_vmcall.cpp

```diff
@@ -0,0 +1,201 @@
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#include <bfjson.h>
+#include <bfstring.h>
+
+#include <exit_handler/exit_handler_intel_x64_eapis.h>
+#include <exit_handler/exit_handler_intel_x64_eapis_vmcall_interface.h>
+
+#include <exit_handler/exit_handler_intel_x64_eapis_verifiers.h>
+#include <exit_handler/exit_handler_intel_x64_eapis_cpuid_verifiers.h>
+
+void
+exit_handler_intel_x64_eapis::register_json_vmcall__cpuid()
+{
+    m_json_commands["emulate_cpuid"] = [&](const auto & ijson, auto & ojson) {
+        this->handle_vmcall__emulate_cpuid(json_hex_or_dec<cpuid_type>(ijson, "leaf"),
+                                           json_hex_or_dec<cpuid_type>(ijson, "subleaf"),
+                                           ijson.at("eax"), ijson.at("ebx"), ijson.at("ecx"), ijson.at("edx"));
+        this->json_success(ojson);
+    };
+
+    m_json_commands["reset_cpuid_leaf"] = [&](const auto & ijson, auto & ojson) {
+        this->handle_vmcall__reset_cpuid_leaf(json_hex_or_dec<cpuid_type>(ijson, "leaf"),
+                                              json_hex_or_dec<cpuid_type>(ijson, "subleaf"));
+        this->json_success(ojson);
+    };
+
+    m_json_commands["reset_cpuid_all"] = [&](const auto &, auto & ojson) {
+        this->handle_vmcall__reset_cpuid_all();
+        this->json_success(ojson);
+    };
+
+    m_json_commands["log_cpuid_access"] = [&](const auto & ijson, auto & ojson){
+        this->handle_vmcall__log_cpuid_access(ijson.at("enabled"));
+        this->json_success(ojson);
+    };
+
+    m_json_commands["clear_cpuid_access_log"] = [&](const auto &, auto & ojson) {
+        this->handle_vmcall__clear_cpuid_access_log();
+        this->json_success(ojson);
+    };
+
+    m_json_commands["cpuid_access_log"] = [&](const auto &, auto & ojson) {
+        this->handle_vmcall__cpuid_access_log(ojson);
+    };
+
+    m_json_commands["dump_cpuid_emulations_log"] = [&](const auto &, auto & ojson) {
+        this->handle_vmcall__dump_cpuid_emulations_log(ojson);
+    };
+}
+
+void
+exit_handler_intel_x64_eapis::handle_vmcall__emulate_cpuid(
+    cpuid_type leaf, cpuid_type subleaf,
+    std::string eax, std::string ebx, std::string ecx, std::string edx)
+{
+    if (policy(emulate_cpuid)->verify(leaf, subleaf, eax, ebx, ecx, edx) != vmcall_verifier::allow) {
+        policy(emulate_cpuid)->deny_vmcall();
+    }
+
+    std::vector<std::string> in_strings = { eax, ebx, ecx, edx };
+    cpuid_regs_type machine_regs = x64::cpuid::get(leaf, 0, subleaf, 0);
+    std::vector<cpuid_key_type> m_regs = { machine_regs.rax, machine_regs.rbx, machine_regs.rcx, machine_regs.rdx };
+
+    for (unsigned long reg_count = 0; reg_count < 4; reg_count++) {
+        std::string str = in_strings.at(reg_count);
+        cpuid_key_type defined_mask = 0;
+        cpuid_key_type passthru_mask = 0;
+        uint64_t num_chars  = str.size() - 1;
```

Comment:
this line to 97 is confusing. Shouldn't size always be 32? If so there is no need for the num_chars or pow variables. Try to clean this up and make the intent clear, maybe even break it up into a couple of functions that state the intent of the code